File: 1396.py
class UndergroundSystem:
    def __init__(self):
        # id -> (startStation, checkInTime)
        self.in_m = {}
        # (start, end) -> [totalTime, numberOfPeople]
        self.t_m = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        if id in self.in_m:
            self.in_m.pop(id)
        self.in_m[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        timeTaken = t - self.in_m[id][1]
        source = self.in_m[id][0]
        if (source, stationName) in self.t_m:
            self.t_m[(source, stationName)][0] += timeTaken
            self.t_m[(source, stationName)][1] += 1
            return
        self.t_m[(source, stationName)] = [timeTaken, 1]

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        i = self.t_m[(startStation, endStation)]
        return i[0] / i[1]


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

File: 1396.cpp
#include <map>
#include <string>

using namespace std;

class UndergroundSystem {
private:
  struct Customer {
    int t;
    string source;
  };
  struct Average {
    long long total;
    long long count;
    double getAverage() {
      return static_cast<double>(total) / static_cast<double>(count);
    }
  };
  string concatenate(string s, string u) { return s + " #\\# " + u; };
  map<int, Customer> cm;
  map<string, Average> tm;

public:
  UndergroundSystem() {}

  void checkIn(int id, string stationName, int t) {
    if (cm.find(id) != cm.end()) {
      cm.erase(id);
    }
    cm[id] = {t, stationName};
  }

  void checkOut(int id, string stationName, int t) {
    string concat = concatenate(cm[id].source, stationName);
    long long timeTaken = t - cm[id].t;
    // if it was already there
    if (tm.find(concat) != tm.end()) {
      ++tm[concat].count;
      tm[concat].total += timeTaken;
      return;
    }
    // if it wasn't already there
    tm[concat] = {timeTaken, 1};
  }

  double getAverageTime(string startStation, string endStation) {
    string concat = concatenate(startStation, endStation);
    return tm[concat].getAverage();
  }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */